### ISSUE TITLE
WARNING: NMS time limit XXXs exceeded for small batch size 8/16 with large images (>960)

### ISSUE THREAD

**Issue by sarmientoj24** [ultralytics/yolov5#7390]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Currently getting

```

     Epoch   gpu_mem       box       obj       cls    labels  img_size
     7/199     11.8G   0.07534    0.1837   0.09136       313      1216: 100%|██████████| 34/34 [00:08<00:00,  3.88it/s]                                         
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95:   0%|          | 0/2 [00:00<?, ?it/s]                                   WARNING: NMS time limit 0.580s exceeded
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95:  50%|█████     | 1/2 [00:00<00:00,  1.37it/s]                           WARNING: NMS time limit 0.520s exceeded
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 2/2 [00:01<00:00,  1.39it/s]                           
                 all         30        754     0.0162      0.109     0.0136    0.00445
```

Which I never got before

**Comment by glenn-jocher** [comment]
@sarmientoj24 default NMS time limits are here, you can modify to whatever you want:
https://github.com/ultralytics/yolov5/blob/4bb7eb8b849fc8a90823a60e2b7a8ec9e38926bf/utils/general.py#L759

**Comment by hellonun** [comment]
@glenn-jocher If it keeps exceeding the time limit in each epoch, will that affect the result of the trained model?
For context, I'm training the yolov5 model on the VisDrone dataset on Colab Tutorial notebook.

**Comment by glenn-jocher** [comment]
@hellonun no, validation has no effect on training.

**Comment by Mronghy** [comment]
> @hellonun no, validation has no effect on training.

Hello, when I am using YOLOv8 to detect videos in track mode (approximately two minutes long), the terminal displays' warning ' ⚠️  Nms time limit 0.550s exceeded ", and then the terminal will display a line of words" killed ". The Yolov8 I deployed on Ubuntu is configured with Intel i7-11700 @ 2.50 GHz and two GeForce RTX 3060 12G graphics cards. Due to the launch of Yolov8 this year, this solution is rarely searchable. May I ask how to solve this problem. The following is what I got when running track:

video 1/1 (2503/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 5.1ms
video 1/1 (2504/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 5.1ms
video 1/1 (2505/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 5.4ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2506/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 350.5ms
video 1/1 (2507/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 6348.7ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2508/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 21653.4ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2509/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 27580.7ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2510/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 33070.3ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2511/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 54215.9ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2512/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 71433.2ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2513/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 85896.8ms
WARNING ⚠️ NMS time limit 0.550s exceeded
video 1/1 (2514/6180) /home/lzz/longda/ultralytics-main/bnz1200/video/1.mp4: 384x640 1 cornea, 1 iris, 1 phacoemulsification head, 61458.9ms
WARNING ⚠️ NMS time limit 0.550s exceeded
killed
(yolo8) lzz@lzz-System-Product-Name:~/longda/ultralytics-main$

**Comment by glenn-jocher** [comment]
@Mronghy It sounds like you're experiencing slow performance due to the complexity of the detection algorithm and the volume of video frames it's processing. One way to improve this issue is to reduce the resolution of your video to a lower size. Alternatively, you can try deploying the model on a powerful machine with faster CPU and GPU. 

Also, take a look at this specific line from your output:
WARNING ⚠️ NMS time limit 0.550s exceeded

That means your current NMS time limit is set to 0.550 seconds, and the calculation took longer than that. You may want to increase the NMS time limit. In the Yolo v5 tracking file, you can see this line of code: 

nms_max_overlap = 1.0  # maximum overlap allowed per class across all boxes

You can change the value for `nms_max_overlap` to whatever value you prefer. Try changing it to a lower value first and see if that helps with the processing speed.

**Comment by Mronghy** [comment]
Thank you for your reply, but I did not find 'nms_max_overlap=1.0' in the Yolov8 project code. But on my other laptop (AMD Ryzen 7 5800H @ 3.20 GHz, NVIDIA Geforce RTX3050ti) using the same video, I was able to run track mode. And I also have a question. The background of the problem is that I have already trained with my own dataset and obtained 'best. pt'. How should I calculate Fps next? When I was in predict mode, I obtained the model testing speed as follows: "Speed: 0.7ms preprocess, 9.0ms inference, 1.6ms postprocess per image at shape (1, 3, 640, 640)". Others' blogs tell us that fps uses 1 second (1000ms) to divide the sum of the above three times. The results I calculated using the speeds of the five networks (n, s, m, l, x) in Yolov8 are not quite the same as those given in the official documents, and the fps of Yolov8x.pt is smaller than that of Yolov8s.pt. Is this due to issues with my personal dataset, or is YOLOV8S performing better than YOLOV8X in small datasets?

**Comment by glenn-jocher** [comment]
@Mronghy sorry about that. In YOLOv5, the `nms_max_overlap` parameter is used in the "detect" file, but not in the "track" file.

Regarding your question on calculating FPS, you are correct that the FPS calculation for YOLOv5 is the inverse of the total processing time for each frame, which is the sum of the preprocess, inference, and postprocess times you mentioned. So you can use the inverse of the sum of those times to get your FPS. 

Regarding your performance discrepancies between the different YOLOv8 models, it's possible that some models perform better than others on specific datasets. It can also depend on the hardware and software configurations of your system. If your dataset is significantly different from the datasets used to train the YOLOv8 models, it's possible that their performance on your dataset may vary. You may need to experiment with different models, configurations, and parameter settings to find the best performance for your specific scenario.

**Comment by Mronghy** [comment]
Thank you again for your reply, which answered the problem I encountered while using YOLO for object detection. I hope YOLO becomes stronger and stronger!!!

**Comment by glenn-jocher** [comment]
@Mronghy you're welcome! We're glad to have been able to help. And yes, we are constantly working to improve YOLO and make it even more powerful for object detection tasks. Thank you for using YOLO and good luck with your project!

**Comment by Mronghy** [comment]
> @Mronghy you're welcome! We're glad to have been able to help. And yes, we are constantly working to improve YOLO and make it even more powerful for object detection tasks. Thank you for using YOLO and good luck with your project!

Hello, I'm here again. May I ask how to set DDP in YOLOv8? I have two graphics cards. However, during operation, it displays "DDP settings: RANK 0, WORLD_SIZE 2, DIVICE cuda: 0"

**Comment by Mronghy** [comment]
> @Mronghy you're welcome! We're glad to have been able to help. And yes, we are constantly working to improve YOLO and make it even more powerful for object detection tasks. Thank you for using YOLO and good luck with your project!

Hello, I'm here again. May I ask how to set up DDP in YOLOv8? I have two graphics cards. However, during the operation, it displays "DDP settings: RANK 0, WORLD_SIZE 2, DIVICE cuda: 0". Perhaps I need to use Docker to achieve distributed operation?

**Comment by glenn-jocher** [comment]
@Mronghy To use Distributed Data Parallel (DDP) with YOLOv8, you can follow these steps:

1. Install the PyTorch module `torch.distributed`.

2. Initialize the DDP environment with `torch.distributed.init_process_group()` function, providing appropriate arguments. For example:

```python
import torch.distributed as dist

dist.init_process_group(backend='nccl', init_method='...', rank=0, world_size=2)
```

Here, 'nccl' is the backend to use, and the `init_method` argument provides the URL to initialize communications. The `rank` argument is the rank of this process, and `world_size` is the total number of processes.

3. After initializing the environment, create a new `DistributedSampler` that shuffles the data among the different processes. You can pass this sampler to your `DataLoader` to load the data in parallel. For example:

```python
train_sampler = torch.utils.data.distributed.DistributedSampler(train_dataset)
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, shuffle=False, sampler=train_sampler, num_workers=8, pin_memory=True)
```

4. Modify your model to use `torch.nn.parallel.DistributedDataParallel`, which replicates the model across multiple GPUs and processes. For example:

```python
device = torch.device("cuda")
model = MyModel().to(device)
model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[torch.cuda.current_device()], output_device=torch.cuda.current_device())
```

You can pass a list of GPU device ids to `device_ids` and set `output_device` to the index of the GPU device that the loss value should be computed on.

5. Finally, add code to save and load checkpoints in the DDP environment. Specifically, each process should save and load its own model checkpoint, and wrap it in `torch.nn.SyncBatchNorm.convert_sync_batchnorm()` to synchronize batch normalization layers. For example:

```python
model_checkpoint_path = 'model.pt'
local_rank = ... # get rank of this process
dist.barrier() # synchronize all processes
if local_rank == 0:
    # only one process should save the checkpoint
    torch.save({'model': model.module.state_dict(), 'optimizer': optimizer.state_dict()}, model_checkpoint_path)
dist.barrier() # synchronize all processes
checkpoint = torch.load(model_checkpoint_path, map
